From: Simon McVittie <smcv@collabora.com>
Date: Thu, 9 Apr 2026 18:45:04 +0100
Subject: run: Cope with an empty runtime

When FlatpakDir runs extra-data helpers in apply_extra_data(),
if the helper is statically linked, it might not need a runtime at all.
For example the helper for openh264 falls into this category.

[smcv: Resolve conflicts with 1.14.x]

Fixes: ac62ebe3 "run: Use O_PATH fds for the runtime and app deploy directories"
Signed-off-by: Simon McVittie <smcv@collabora.com>
(cherry picked from commit aa1a54c9dae25fd13ebc936e06996f8db39f4aa5)
Origin: backport, 1.16.6, commit:b942152bde980cde51e933359e4ba3be527ae3ed
Bug: https://github.com/flatpak/flatpak/issues/6583
---
 common/flatpak-run.c | 30 +++++++++++++++++++++++-------
 1 file changed, 23 insertions(+), 7 deletions(-)

diff --git a/common/flatpak-run.c b/common/flatpak-run.c
index 0f3d122..0a83d9b 100644
--- a/common/flatpak-run.c
+++ b/common/flatpak-run.c
@@ -2917,6 +2917,10 @@ flatpak_run_add_app_info_args (FlatpakBwrap       *bwrap,
   return TRUE;
 }
 
+/*
+ * @runtime_fd: the /usr for the runtime, or -1 if running with no runtime,
+ *  perhaps to unpack extra-data
+ */
 static void
 add_tzdata_args (FlatpakBwrap *bwrap,
                  int           runtime_fd)
@@ -2929,6 +2933,8 @@ add_tzdata_args (FlatpakBwrap *bwrap,
   glnx_autofd int zoneinfo_fd = -1;
   g_autoptr(GError) error = NULL;
 
+  g_return_if_fail (runtime_fd >= -1);
+
   raw_timezone = flatpak_get_timezone ();
   timezone_content = g_strdup_printf ("%s\n", raw_timezone);
   localtime_content = g_strconcat ("../usr/share/zoneinfo/", raw_timezone, NULL);
@@ -2937,10 +2943,11 @@ add_tzdata_args (FlatpakBwrap *bwrap,
 
   tzdir_fd = glnx_chaseat (AT_FDCWD, tzdir, GLNX_CHASE_MUST_BE_DIRECTORY, NULL);
 
-  zoneinfo_fd = glnx_chaseat (runtime_fd, "share/zoneinfo",
-                              GLNX_CHASE_RESOLVE_BENEATH |
-                              GLNX_CHASE_MUST_BE_DIRECTORY,
-                              NULL);
+  if (runtime_fd >= 0)
+    zoneinfo_fd = glnx_chaseat (runtime_fd, "share/zoneinfo",
+                                GLNX_CHASE_RESOLVE_BENEATH |
+                                GLNX_CHASE_MUST_BE_DIRECTORY,
+                                NULL);
 
   /* Check for host /usr/share/zoneinfo */
   if (tzdir_fd >= 0 && zoneinfo_fd >= 0)
@@ -2951,7 +2958,7 @@ add_tzdata_args (FlatpakBwrap *bwrap,
                               "--symlink", localtime_content, "/etc/localtime",
                               NULL);
     }
-  else
+  else if (runtime_fd >= 0)
     {
       g_autofree char *runtime_zoneinfo = NULL;
       glnx_autofd int runtime_zoneinfo_fd = -1;
@@ -3448,6 +3455,10 @@ setup_seccomp (FlatpakBwrap   *bwrap,
 }
 #endif
 
+/*
+ * @runtime_fd: the /usr for the runtime, or -1 if running with no runtime,
+ *  perhaps to unpack extra-data
+ */
 static void
 flatpak_run_setup_usr_links (FlatpakBwrap *bwrap,
                              int          runtime_fd,
@@ -3501,6 +3512,10 @@ flatpak_run_setup_usr_links (FlatpakBwrap *bwrap,
     }
 }
 
+/*
+ * @runtime_fd: the /usr for the runtime, or -1 if running with no runtime,
+ *  perhaps to unpack extra-data
+ */
 gboolean
 flatpak_run_setup_base_argv (FlatpakBwrap   *bwrap,
                              int             runtime_fd,
@@ -3517,7 +3532,7 @@ flatpak_run_setup_base_argv (FlatpakBwrap   *bwrap,
   gulong pers;
   gid_t gid = getgid ();
 
-  g_return_val_if_fail (runtime_fd >= 0, FALSE);
+  g_return_val_if_fail (runtime_fd >= -1, FALSE);
 
   run_dir = g_strdup_printf ("/run/user/%d", getuid ());
 
@@ -3592,7 +3607,8 @@ flatpak_run_setup_base_argv (FlatpakBwrap   *bwrap,
   else if (g_file_test ("/var/lib/dbus/machine-id", G_FILE_TEST_EXISTS))
     flatpak_bwrap_add_args (bwrap, "--ro-bind", "/var/lib/dbus/machine-id", "/etc/machine-id", NULL);
 
-  if ((flags & FLATPAK_RUN_FLAG_WRITABLE_ETC) == 0)
+  if (runtime_fd >= 0
+      && (flags & FLATPAK_RUN_FLAG_WRITABLE_ETC) == 0)
     {
       g_auto(GLnxDirFdIterator) dfd_iter = { 0, };
       struct dirent *dent;
